### PR TITLE
issue #373. write modified lfric kernels

### PR DESCRIPTION
**Pull request by rupertford**

### CODE REVIEW

**Comment by rupertford** [comment]
Ready for review by @arporter or @sergisiso

File: src/psyclone/tests/kernel_transformation_test.py [hunk 1]
```diff
@@ -42,7 +42,8 @@
 from fparser.two.utils import walk_ast
 from dynamo0p3_build import Dynamo0p3Build
 from psyclone_test_utils import get_invoke
-from psyclone.transformations import TransformationError, ACCRoutineTrans
+from psyclone.transformations import TransformationError, ACCRoutineTrans, \
+    Dynamo0p3KernelConstTrans
```
Comment (arporter) [inline src/psyclone/tests/kernel_transformation_test.py]: Please document the Dynamo0p3KernelConstTrans transformation in section 7.3.2 of the User Guide.

File: src/psyclone/tests/psyir/backend/fortran_test.py [hunk 2]
```diff
@@ -637,3 +637,42 @@ def test_fw_codeblock():
         "    a=1\n"
         "PRINT *, 'I am a code block'\n"
         "    PRINT *, 'with more than one line'\n" in result)
+
+
+@pytest.mark.xfail(reason="issue #430 : module name should be specified")
+def test_module_name(monkeypatch):
```
Comment (arporter) [inline src/psyclone/tests/psyir/backend/fortran_test.py]: `monkeypatch` is not used and pylint complains about a missing import.

File: src/psyclone/psyGen.py [hunk 3]
```diff
@@ -3823,9 +3833,26 @@ def rename_and_write(self):
...
+            new_kern_code = fortran_writer(self.get_kernel_schedule())
+            fll = FortLineLength()
+            new_kern_code = fll.process(new_kern_code)
         else:
-            # Generate the Fortran for this transformed kernel, ensuring that
-            # we limit the line lengths
+            # This is an old style transformation which modifes the
+            # fp2 parse tree directly. Therefore use the fp2
+            # representation to generate the Fortran for this
+            # transformed kernel, ensuring that limit the line length
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/psyGen.py]: Done.

File: src/psyclone/psyGen.py [hunk 4]
```diff
@@ -3855,6 +3882,37 @@ def rename_and_write(self):
             # Close the new kernel file
             os.close(fdesc)
 
+    def _rename_psyir(self, suffix):
+        '''Rename the PSyIR module and kernel names by adding the supplied
+        suffix to the names. This change affects the KernCall and
+        KernelSchedule nodes. Currently it is only possible to set the
+        kernel name in a KernCall node. The kernel module is then
+        inferred from this by assuming a fixxed connection between
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/psyGen.py]: fixed fixxed.
No the assumption was not checked. I've added a check and associated test.

File: src/psyclone/tests/psyGen_test.py [hunk 5]
```diff
@@ -5305,3 +5304,29 @@ def test_fparser2astprocessor_handling_end_subroutine_stmt(f2008_parser):
...
+
+
+def test_modified_kern_line_length(tmpdir, monkeypatch):
+    '''Modified Fortran kernels are written to file linewrapped at 132
+    characters. This test checks that this linewrapping works.
+
+    '''
+    from psyclone.transformations import Dynamo0p3KernelConstTrans
+    # Ensure kernel-output directory is uninitialised
+    config = Config.get()
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/psyGen_test.py]: I've moved the fixture and updated relevant tests.

I agree that fixtures should be documented but don't think its the job of this pr. Perhaps raise an issue?

File: src/psyclone/psyir/backend/fortran.py [hunk 6]
```diff
@@ -221,9 +221,10 @@ def kernelschedule_node(self, node):
         if not node.name:
             raise VisitorError("Expected node name to have a value.")
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/psyir/backend/fortran.py]: Well spotted. Added.

File: src/psyclone/psyir/backend/fortran.py [hunk 7]
```diff
@@ -221,9 +221,10 @@ def kernelschedule_node(self, node):
...
 
+        module_name = node.name.rstrip("_code") + "_mod"
         result = (
             "{0}module {1}\n"
-            "".format(self._nindent, node.name+"_mod"))
+            "".format(self._nindent, module_name))
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/psyir/backend/fortran.py]: It's at the end of the method and was potentially inconsistent so I've made it consistent.

File: examples/dynamo/eg13/kernel_constants.py [hunk 8]
```diff
@@ -79,7 +78,6 @@ def trans(psy):
```
[the remark this answers is not in the document]
Comment (rupertford) [reply examples/dynamo/eg13/kernel_constants.py]: Added some words.

File: src/psyclone/tests/kernel_transformation_test.py (hunk 1 above)
Comment (rupertford) [reply src/psyclone/tests/kernel_transformation_test.py]: Done (I think).

File: src/psyclone/tests/psyir/backend/fortran_test.py (hunk 2 above)
Comment (rupertford) [reply src/psyclone/tests/psyir/backend/fortran_test.py]: Well observed.

**Comment by rupertford** [comment]
Ready for next review from @arporter (assuming the tests work).

File: src/psyclone/tests/kernel_transformation_test.py [hunk 9]
```diff
@@ -186,15 +184,10 @@ def test_new_kernel_file(tmpdir, monkeypatch):
     # If compilation fails this will raise an exception
     GOcean1p0Build(tmpdir).compile_file(filename)
 
-    old_cwd.chdir()
 
-
-def test_new_kernel_dir(tmpdir, monkeypatch):
+def test_new_kernel_dir(outputdir, tmpdir, monkeypatch):
```
Comment (arporter) [inline src/psyclone/tests/kernel_transformation_test.py]: Again, I think we can get rid of tmpdir now.

File: src/psyclone/tests/kernel_transformation_test.py [hunk 10]
```diff
@@ -231,18 +221,73 @@ def test_new_kern_no_clobber(tmpdir, monkeypatch):
     _ = str(psy.gen).lower()
     filename = os.path.join(str(tmpdir), old_mod_name+"_1_mod.f90")
```
Comment (arporter) [inline src/psyclone/tests/kernel_transformation_test.py]: Can we replace tmpdir with outputdir?

File: src/psyclone/tests/kernel_transformation_test.py [hunk 11]
```diff
@@ -231,18 +221,73 @@ def test_new_kern_no_clobber(tmpdir, monkeypatch):
...
+    sched = invoke.schedule
+    kernels = sched.walk(sched.children, Kern)
+    kern = kernels[0]
+    ktrans = Dynamo0p3KernelConstTrans()
+    _, _ = ktrans.apply(kern, number_of_layers=100)
+    monkeypatch.setattr(kern, "_module_name", mod_name)
+    monkeypatch.setattr(kern, "_name", sub_name)
+    # Generate the code - this should not raise an exception.
+    kern.rename_and_write()
+    filename = os.path.join(str(tmpdir), mod_name[:8]+"_0_mod.f90")
```
Comment (arporter) [inline src/psyclone/tests/kernel_transformation_test.py]: Again, can we replace tmpdir with outputdir?

File: src/psyclone/tests/kernel_transformation_test.py [hunk 12]
```diff
@@ -231,18 +221,73 @@ def test_new_kern_no_clobber(tmpdir, monkeypatch):
...
+    assert os.path.isfile(filename)
+
+
+def test_new_kern_single_error(outputdir, tmpdir, monkeypatch):
```
Comment (arporter) [inline src/psyclone/tests/kernel_transformation_test.py]: Drop tmpdir?

File: src/psyclone/tests/kernel_transformation_test.py [hunk 13]
```diff
@@ -263,19 +308,15 @@ def test_new_kern_single_error(tmpdir, monkeypatch):
             "exists in the kernel-output directory ({0}) but is not the same "
             "as the current, transformed kernel and the kernel-renaming "
             "scheme is set to 'single'".format(str(tmpdir)) in str(err))
-    old_cwd.chdir()
 
 
-def test_new_same_kern_single(tmpdir, monkeypatch):
+def test_new_same_kern_single(outputdir, tmpdir, monkeypatch):
```
Comment (arporter) [inline src/psyclone/tests/kernel_transformation_test.py]: Drop tmpdir and throughout.

File: src/psyclone/tests/psyGen_test.py [hunk 14]
```diff
@@ -5163,20 +5163,20 @@ def test_fparser2astprocessor_handling_end_subroutine_stmt(f2008_parser):
     assert not fake_parent.children  # No new children created
 
 
-def test_modified_kern_line_length(tmpdir, monkeypatch):
+def test_modified_kern_line_length(outputdir, monkeypatch, tmpdir):
```
Comment (arporter) [inline src/psyclone/tests/psyGen_test.py]: Drop tmpdir.

File: src/psyclone/tests/psyir/backend/fortran_test.py [hunk 15]
```diff
@@ -640,7 +640,7 @@ def test_fw_codeblock():
 
 
 @pytest.mark.xfail(reason="issue #430 : module name should be specified")
-def test_module_name(monkeypatch):
```
Comment (arporter) [inline src/psyclone/tests/psyir/backend/fortran_test.py]: pylint complains:

    fortran_test.py:42:0: C0414: Import alias does not rename original package (useless-import-alias)

File: src/psyclone/tests/kernel_transformation_test.py [hunk 16]
```diff
@@ -136,16 +137,13 @@ def test_accroutine_empty_kernel():
     assert "!$acc routine\n  end subroutine testkern_code" in gen
 
 
-def test_new_kernel_file(tmpdir, monkeypatch):
+def test_new_kernel_file(outputdir, tmpdir, monkeypatch):
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/kernel_transformation_test.py]: Also, pylint complains that a lot of `monkeypatch`s are not used.

File: src/psyclone/tests/psyir/backend/fortran_test.py [hunk 17]
```diff
@@ -637,3 +637,42 @@ def test_fw_codeblock():
         "    a=1\n"
         "PRINT *, 'I am a code block'\n"
         "    PRINT *, 'with more than one line'\n" in result)
+
```
Comment (arporter) [inline src/psyclone/tests/psyir/backend/fortran_test.py]: pylint complains about a missing absolute_import and an Import alias not renaming original package. Neither of those are the fault of this PR though.

File: doc/user_guide/dynamo0p3.rst [hunk 18]
```diff
@@ -2109,6 +2109,10 @@ that supports distributed memory.  An example of redundant computation
 can be found in ``examples/dynamo/eg8`` and an example of asynchronous
 halo exchanges can be found in ``examples/dynamo/eg11``.
 
+The **Dynamo0p3KernelConstTrans** transformation is only valid for the
+"Dynamo0p3" API. This is because the properties that it makes constant
+are specific to the "Dynamo0p3" API
```
[the remark this answers is not in the document]
Comment (rupertford) [reply doc/user_guide/dynamo0p3.rst]: Better thanks.

File: doc/user_guide/dynamo0p3.rst [hunk 19]
```diff
@@ -2140,6 +2144,10 @@ all versions of the Dynamo API.
     :members:
     :noindex:
 
+.. autoclass:: psyclone.transformations.Dynamo0p3KernelConstTrans
```
[the remark this answers is not in the document]
Comment (rupertford) [reply doc/user_guide/dynamo0p3.rst]: Oops, I should have checked that.

File: doc/user_guide/examples.rst [hunk 20]
```diff
@@ -190,6 +190,17 @@ to the specified Kernel:
 
   > python find_kernel.py
 
+Example 13
```
[the remark this answers is not in the document]
Comment (rupertford) [reply doc/user_guide/examples.rst]: OK, done this, but the previous 12 examples don't have a sub-title.

File: src/psyclone/tests/conftest.py [hunk 21]
```diff
@@ -140,3 +141,11 @@ def parser():
     '''
     from fparser.two.parser import ParserFactory
     return ParserFactory().create()
+
+
+@pytest.fixture(scope="function")
+def outputdir(tmpdir, monkeypatch):
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/conftest.py]: OK, changed them all to kernel_outputdir. I also found a gocean test file which specified this fixture so I removed it.

File: src/psyclone/tests/gocean1p0_opencl_test.py [hunk 22]
```diff
@@ -32,6 +32,7 @@
 # POSSIBILITY OF SUCH DAMAGE.
 # ----------------------------------------------------------------------------
 # Authors A. R. Porter and S. Siso, STFC Daresbury Lab
+# Modified by R. W.  Siso, STFC Daresbury Lab.
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/gocean1p0_opencl_test.py]: Oops :-)

File: src/psyclone/tests/kernel_transformation_test.py (hunk 12 above)
Comment (rupertford) [reply src/psyclone/tests/kernel_transformation_test.py]: Removed all tmpdir and all unused monkeypatch.

File: src/psyclone/tests/kernel_transformation_test.py (hunk 13 above)
Comment (rupertford) [reply src/psyclone/tests/kernel_transformation_test.py]: Done, see above.

File: src/psyclone/tests/psyGen_test.py (hunk 14 above)
Comment (rupertford) [reply src/psyclone/tests/psyGen_test.py]: Done.

File: src/psyclone/tests/psyir/backend/fortran_test.py (hunk 15 above)
Comment (rupertford) [reply src/psyclone/tests/psyir/backend/fortran_test.py]: I can't get my pylint (1.9.5) to complain, but I've modified the code in any case.

File: src/psyclone/tests/psyir/backend/fortran_test.py (hunk 17 above)
Comment (rupertford) [reply src/psyclone/tests/psyir/backend/fortran_test.py]: Again I don't see these in either pylint 1.9.5 (python2) or pylint 2.3.1 (python3). I've added an absolute import and will hope the renaming problem is the same issue as the previous comment.

**Comment by rupertford** [comment]
I think this is ready for the next review (from @arporter)

File: doc/user_guide/examples.rst (hunk 20 above)
[the remark this answers is not in the document]
Comment (arporter) [reply doc/user_guide/examples.rst]: True but I think it helps the reader and this is the only one modified in this PR :-)

File: src/psyclone/tests/psyir/backend/fortran_test.py (hunk 15 above)
Comment (arporter) [reply src/psyclone/tests/psyir/backend/fortran_test.py]: You've fixed it. FWIW, my pylint is 2.1.1.

File: src/psyclone/tests/psyir/backend/fortran_test.py (hunk 17 above)
Comment (arporter) [reply src/psyclone/tests/psyir/backend/fortran_test.py]: Code now scores 10.0 :-)

**Comment by arporter** [comment]
I was all set to merge this but one test fails (kernel_transformation_test::test_2kern_trans) when compilation is enabled.

    Error: Dummy argument scalar with INTENT(IN) in variable definition context (actual argument to INTENT = OUT/INOUT) at (1) psy.f90:203:129:

Once that is fixed this can be merged.

**Comment by rupertford** [comment]
Thanks for checking with --compile. I should have done that myself. The problem is not to do with the changes in this PR. The problem is that when we translate kernels from Fortran to the PSyIR, any argument with a missing intent is given the value inout. This was causing the test to fail as the metadata declared the argument as READ but no intent was specified. It was agreed at the telco that LFRic should always provide an intent for kernels. I've created issue #462 for us to check this. I think that in general we should leave the intent blank if we do not know what it is, or perhaps use the metadata to ensure it is consistent.

**Comment by rupertford** [comment]
Ready for next review from @arporter

**Comment by arporter** [comment]
Tests all now pass with compilation enabled. Please fix the conflict with master and then this PR can be merged.

**Comment by rupertford** [comment]
Ready for next review from @arporter, subject to successful travis checks.

**Comment by arporter** [comment]
Rather than bounce this again I've taken the liberty of fixing the new conflicts (since they were very simple). Tests all pass with compilation enabled.

**Pull request merged**